// angular-material-extensions/esm2015/table/directives/custom-header.directive.js
import { Directive, ElementRef } from '@angular/core';
import { DragDrop, moveItemInArray } from '@angular/cdk/drag-drop';
import { MatTable } from '@angular/material/table';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/material/table';
import * as ɵngcc2 from '@angular/cdk/drag-drop';
export class CustomHeaderDirective {
    constructor(table, headerRow, dndService) {
        this.table = table;
        this.headerRow = headerRow;
        this.dndService = dndService;
        this.columns = [];
    }
    ngAfterViewInit() {
        this.dropListRef = this.createDropList(this.headerRow, this.createDrags(this.getHeaderCells(this.headerRow.nativeElement)));
        this.columns = this.table._contentHeaderRowDefs.first.columns;
        console.log(this.columns);
    }
    drop(event) {
        moveItemInArray(this.columns, event.previousIndex, event.currentIndex);
    }
    getHeaderCells(rowElem) {
        return Array.from(rowElem.children);
    }
    createDrags(cellRefs) {
        return cellRefs.map((cellRef) => {
            const drag = this.dndService.createDrag(cellRef);
            drag.lockAxis = 'x';
            return drag;
        });
    }
    createDropList(headerRowDef, drags) {
        const dropListRef = this.dndService.createDropList(headerRowDef);
        dropListRef.withItems(drags);
        dropListRef.withOrientation('horizontal');
        dropListRef.dropped.subscribe((event) => this.drop(event));
        return dropListRef;
    }
}
CustomHeaderDirective.ɵfac = function CustomHeaderDirective_Factory(t) { return new (t || CustomHeaderDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.MatTable), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.DragDrop)); };
CustomHeaderDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: CustomHeaderDirective, selectors: [["", "matCustomHeader", ""]] });
CustomHeaderDirective.ctorParameters = () => [
    { type: MatTable },
    { type: ElementRef },
    { type: DragDrop }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CustomHeaderDirective, [{
        type: Directive,
        args: [{
                selector: '[matCustomHeader]'
            }]
    }], function () { return [{ type: ɵngcc1.MatTable }, { type: ɵngcc0.ElementRef }, { type: ɵngcc2.DragDrop }]; }, null); })();
