// angular-material-extensions/esm2015/table/table.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatTableModule } from '@angular/material/table';
import { MatCheckboxModule } from '@angular/material/checkbox';
import { CustomHeaderDirective } from './directives/custom-header.directive';
import { MatSelectionColumnComponent } from './selection-column/selection-column.component';
/** @description Exports all the extensions for material's table component. */
import * as ɵngcc0 from '@angular/core';
export class MatTableExtensionsModule {
}
MatTableExtensionsModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: MatTableExtensionsModule });
MatTableExtensionsModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function MatTableExtensionsModule_Factory(t) { return new (t || MatTableExtensionsModule)(); }, imports: [[
            CommonModule,
            MatTableModule,
            MatCheckboxModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MatTableExtensionsModule, { declarations: function () { return [CustomHeaderDirective,
        MatSelectionColumnComponent]; }, imports: function () { return [CommonModule,
        MatTableModule,
        MatCheckboxModule]; }, exports: function () { return [CustomHeaderDirective,
        MatSelectionColumnComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatTableExtensionsModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    CustomHeaderDirective,
                    MatSelectionColumnComponent
                ],
                imports: [
                    CommonModule,
                    MatTableModule,
                    MatCheckboxModule
                ],
                exports: [
                    CustomHeaderDirective,
                    MatSelectionColumnComponent
                ]
            }]
    }], null, null); })();
