// angular-material-extensions/bundles/angular-material-extensions.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/material/table'),require('@angular/cdk/drag-drop'),require('@angular/material/checkbox'),require('@angular/common'),exports, require('@angular/core'), require('@angular/common'), require('@angular/material/table'), require('@angular/material/checkbox'), require('@angular/cdk/drag-drop')) :
    typeof define === 'function' && define.amd ? define('angular-material-extensions', ['@angular/core','@angular/material/table','@angular/cdk/drag-drop','@angular/material/checkbox','@angular/common','exports', '@angular/core', '@angular/common', '@angular/material/table', '@angular/material/checkbox', '@angular/cdk/drag-drop'], factory) :
    (global = global || self, factory(global.ng.core,global.ng.material.table,global.ng.cdk.dragDrop,global.ng.material.checkbox,global.ng.common,global['angular-material-extensions'] = {}, global.ng.core, global.ng.common, global.ng.material.table, global.ng.material.checkbox, global.ng.cdk.dragDrop));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,ɵngcc4,exports, core, common, table, checkbox, dragDrop) { 
function MatSelectionColumnComponent_th_1_Template(rf, ctx) { if (rf & 1) {
    var _r3 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "th", 3);
    ɵngcc0.ɵɵelementStart(1, "mat-checkbox", 4);
    ɵngcc0.ɵɵlistener("change", function MatSelectionColumnComponent_th_1_Template_mat_checkbox_change_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r3); var ctx_r2 = ɵngcc0.ɵɵnextContext(); return $event ? ctx_r2.masterToggle() : null; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("checked", (ctx_r0.selection == null ? null : ctx_r0.selection.hasValue()) && ctx_r0.isAllSelected())("indeterminate", (ctx_r0.selection == null ? null : ctx_r0.selection.hasValue()) && !ctx_r0.isAllSelected());
} }
function MatSelectionColumnComponent_td_2_Template(rf, ctx) { if (rf & 1) {
    var _r6 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "td", 5);
    ɵngcc0.ɵɵelementStart(1, "mat-checkbox", 6);
    ɵngcc0.ɵɵlistener("click", function MatSelectionColumnComponent_td_2_Template_mat_checkbox_click_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); return $event.stopPropagation(); })("change", function MatSelectionColumnComponent_td_2_Template_mat_checkbox_change_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r6); var row_r4 = ctx.$implicit; var ctx_r7 = ɵngcc0.ɵɵnextContext(); return $event ? ctx_r7.selection == null ? null : ctx_r7.selection.toggle(row_r4) : null; });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var row_r4 = ctx.$implicit;
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("checked", ctx_r1.selection == null ? null : ctx_r1.selection.isSelected(row_r4));
} }
'use strict';

    var CustomHeaderDirective = /** @class */ (function () {
        function CustomHeaderDirective(table, headerRow, dndService) {
            this.table = table;
            this.headerRow = headerRow;
            this.dndService = dndService;
            this.columns = [];
        }
        CustomHeaderDirective.prototype.ngAfterViewInit = function () {
            this.dropListRef = this.createDropList(this.headerRow, this.createDrags(this.getHeaderCells(this.headerRow.nativeElement)));
            this.columns = this.table._contentHeaderRowDefs.first.columns;
            console.log(this.columns);
        };
        CustomHeaderDirective.prototype.drop = function (event) {
            dragDrop.moveItemInArray(this.columns, event.previousIndex, event.currentIndex);
        };
        CustomHeaderDirective.prototype.getHeaderCells = function (rowElem) {
            return Array.from(rowElem.children);
        };
        CustomHeaderDirective.prototype.createDrags = function (cellRefs) {
            var _this = this;
            return cellRefs.map(function (cellRef) {
                var drag = _this.dndService.createDrag(cellRef);
                drag.lockAxis = 'x';
                return drag;
            });
        };
        CustomHeaderDirective.prototype.createDropList = function (headerRowDef, drags) {
            var _this = this;
            var dropListRef = this.dndService.createDropList(headerRowDef);
            dropListRef.withItems(drags);
            dropListRef.withOrientation('horizontal');
            dropListRef.dropped.subscribe(function (event) { return _this.drop(event); });
            return dropListRef;
        };
CustomHeaderDirective.ɵfac = function CustomHeaderDirective_Factory(t) { return new (t || CustomHeaderDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.MatTable), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.DragDrop)); };
CustomHeaderDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: CustomHeaderDirective, selectors: [["", "matCustomHeader", ""]] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CustomHeaderDirective, [{
        type: core.Directive,
        args: [{
                selector: '[matCustomHeader]'
            }]
    }], function () { return [{ type: ɵngcc1.MatTable }, { type: ɵngcc0.ElementRef }, { type: ɵngcc2.DragDrop }]; }, null); })();
        return CustomHeaderDirective;
    }());
    CustomHeaderDirective.ctorParameters = function () { return [
        { type: table.MatTable },
        { type: core.ElementRef },
        { type: dragDrop.DragDrop }
    ]; };

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, exports) {
        for (var p in m)
            if (p !== "default" && !exports.hasOwnProperty(p))
                __createBinding(exports, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (Object.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var MatSelectionColumnComponent = /** @class */ (function () {
        function MatSelectionColumnComponent(table) {
            this.table = table;
        }
        MatSelectionColumnComponent.prototype.ngOnInit = function () {
            this.columnDef.name = this.name || 'select';
            this.columnDef.cell = this.cell;
            this.columnDef.headerCell = this.headerCell;
            this.table.addColumnDef(this.columnDef);
        };
        MatSelectionColumnComponent.prototype.ngOnDestroy = function () {
            this.table.removeColumnDef(this.columnDef);
        };
        Object.defineProperty(MatSelectionColumnComponent.prototype, "data", {
            get: function () {
                if (!this.table) {
                    return [];
                }
                if (this.table.dataSource instanceof table.MatTableDataSource) {
                    return this.table.dataSource.data || [];
                }
                // TODO: What about Observables?
                return this.table.dataSource || [];
            },
            enumerable: false,
            configurable: true
        });
        /** Whether the number of selected elements matches the total number of rows. */
        MatSelectionColumnComponent.prototype.isAllSelected = function () {
            var _a;
            var numSelected = ((_a = this.selection) === null || _a === void 0 ? void 0 : _a.selected.length) || 0;
            var numRows = this.data.length;
            return numSelected === numRows;
        };
        /** Selects all rows if they are not all selected; otherwise clear selection. */
        MatSelectionColumnComponent.prototype.masterToggle = function () {
            var _b;
            if (!this.selection) {
                return;
            }
            this.isAllSelected() ?
                this.selection.clear() : (_b = this.selection).select.apply(_b, __spread(this.data));
        };
MatSelectionColumnComponent.ɵfac = function MatSelectionColumnComponent_Factory(t) { return new (t || MatSelectionColumnComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.MatTable)); };
MatSelectionColumnComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatSelectionColumnComponent, selectors: [["mat-selection-column"]], viewQuery: function MatSelectionColumnComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(table.MatColumnDef, true);
        ɵngcc0.ɵɵstaticViewQuery(table.MatCellDef, true);
        ɵngcc0.ɵɵstaticViewQuery(table.MatHeaderCellDef, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.columnDef = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.cell = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.headerCell = _t.first);
    } }, inputs: { name: "name", selection: "selection" }, decls: 3, vars: 0, consts: [["matColumnDef", ""], ["mat-header-cell", "", 4, "matHeaderCellDef"], ["mat-cell", "", 4, "matCellDef"], ["mat-header-cell", ""], [3, "checked", "indeterminate", "change"], ["mat-cell", ""], [3, "checked", "click", "change"]], template: function MatSelectionColumnComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementContainerStart(0, 0);
        ɵngcc0.ɵɵtemplate(1, MatSelectionColumnComponent_th_1_Template, 2, 2, "th", 1);
        ɵngcc0.ɵɵtemplate(2, MatSelectionColumnComponent_td_2_Template, 2, 1, "td", 2);
        ɵngcc0.ɵɵelementContainerEnd();
    } }, directives: [ɵngcc1.MatColumnDef, ɵngcc1.MatHeaderCellDef, ɵngcc1.MatCellDef, ɵngcc1.MatHeaderCell, ɵngcc3.MatCheckbox, ɵngcc1.MatCell], styles: [".mat-cell[_ngcontent-%COMP%], .mat-header-cell[_ngcontent-%COMP%]{padding-left:24px;padding-right:24px}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatSelectionColumnComponent, [{
        type: core.Component,
        args: [{
                selector: 'mat-selection-column',
                template: "<ng-container matColumnDef>\n  <th mat-header-cell *matHeaderCellDef>\n    <mat-checkbox (change)=\"$event ? masterToggle() : null\"\n                  [checked]=\"selection?.hasValue() && isAllSelected()\"\n                  [indeterminate]=\"selection?.hasValue() && !isAllSelected()\">\n    </mat-checkbox>\n  </th>\n  <td mat-cell *matCellDef=\"let row\">\n    <mat-checkbox (click)=\"$event.stopPropagation()\"\n                  (change)=\"$event ? selection?.toggle(row) : null\"\n                  [checked]=\"selection?.isSelected(row)\">\n    </mat-checkbox>\n  </td>\n</ng-container>\n",
                styles: [".mat-cell,.mat-header-cell{padding-left:24px;padding-right:24px}"]
            }]
    }], function () { return [{ type: ɵngcc1.MatTable }]; }, { name: [{
            type: core.Input
        }], selection: [{
            type: core.Input
        }], columnDef: [{
            type: core.ViewChild,
            args: [table.MatColumnDef, { static: true }]
        }], cell: [{
            type: core.ViewChild,
            args: [table.MatCellDef, { static: true }]
        }], headerCell: [{
            type: core.ViewChild,
            args: [table.MatHeaderCellDef, { static: true }]
        }] }); })();
        return MatSelectionColumnComponent;
    }());
    MatSelectionColumnComponent.ctorParameters = function () { return [
        { type: table.MatTable }
    ]; };
    MatSelectionColumnComponent.propDecorators = {
        name: [{ type: core.Input }],
        selection: [{ type: core.Input }],
        columnDef: [{ type: core.ViewChild, args: [table.MatColumnDef, { static: true },] }],
        cell: [{ type: core.ViewChild, args: [table.MatCellDef, { static: true },] }],
        headerCell: [{ type: core.ViewChild, args: [table.MatHeaderCellDef, { static: true },] }]
    };

    /** @description Exports all the extensions for material's table component. */
    var MatTableExtensionsModule = /** @class */ (function () {
        function MatTableExtensionsModule() {
        }
MatTableExtensionsModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: MatTableExtensionsModule });
MatTableExtensionsModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function MatTableExtensionsModule_Factory(t) { return new (t || MatTableExtensionsModule)(); }, imports: [[
            common.CommonModule,
            table.MatTableModule,
            checkbox.MatCheckboxModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MatTableExtensionsModule, { declarations: [CustomHeaderDirective, MatSelectionColumnComponent], imports: [ɵngcc4.CommonModule, ɵngcc1.MatTableModule, ɵngcc3.MatCheckboxModule], exports: [CustomHeaderDirective, MatSelectionColumnComponent] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatTableExtensionsModule, [{
        type: core.NgModule,
        args: [{
                declarations: [
                    CustomHeaderDirective,
                    MatSelectionColumnComponent
                ],
                imports: [
                    common.CommonModule,
                    table.MatTableModule,
                    checkbox.MatCheckboxModule
                ],
                exports: [
                    CustomHeaderDirective,
                    MatSelectionColumnComponent
                ]
            }]
    }], function () { return []; }, null); })();
        return MatTableExtensionsModule;
    }());

    // module

    /*
     * Public API Surface of extensions
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.CustomHeaderDirective = CustomHeaderDirective;
    exports.MatSelectionColumnComponent = MatSelectionColumnComponent;
    exports.MatTableExtensionsModule = MatTableExtensionsModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
